### PR TITLE
Assorted Linux fixes

### PR DESCRIPTION
**Pull request by digitalinfinity**
- Turn off fpo- stack walking depends on it on retail builds
- Speed up debug builds by improving performance of IsAddressOnStack
- Add support for debug and retail builds being in different directories

### CODE REVIEW

File: pal/src/thread/thread.cpp [hunk 1]
```diff
@@ -2820,3 +2820,32 @@ void GetCurrentThreadStackLimits(ULONG_PTR* lowLimit, ULONG_PTR* highLimit)
     *highLimit = (ULONG_PTR) stackbase;
 #endif
 }
+
+static thread_local ULONG_PTR s_cachedThreadHighLimit = 0;
```
Comment (jianchun) [inline pal/src/thread/thread.cpp]: nit pick: `s_cachedThreadHighLimit` variable name doesn't indicate it's stack related.

**Comment by jianchun** [comment]
Changing binary drop location may require runtests.py update?

LGTM otherwise.

**Comment by obastemur** [comment]
> Speed up debug builds by improving performance of IsAddressOnStack

Curious, how did we do that?

'fno-omit-frame-pointer' does otherwise.

**Comment by digitalinfinity** [comment]
@obastemur the two are inrelated- -fno-omit-frame-pointer should not noticeably affect build time (it might slow down *runtime* on functions with high register pressure). The debug build speed up was done by improving the performance of IsAddressOnStack- it used to call GetThreadStackBounds very frequently and the Linux implementation of that function is slow- I optimized IsAddressOnStack by not having to call GetThreadStackBounds and can do a much cheaper check.

@jianchun looks like runtests.sh needs to be updated too- I'll update the PR

File: pal/src/thread/thread.cpp [hunk 2]
```diff
@@ -2820,3 +2820,32 @@ void GetCurrentThreadStackLimits(ULONG_PTR* lowLimit, ULONG_PTR* highLimit)
...
+
+bool IsAddressOnStack(ULONG_PTR address)
```
Comment (obastemur) [inline pal/src/thread/thread.cpp]: Should it be better to check if there is an upper limit in case of `ulimit -s unlimited` ? `getrlimit(RLIMIT_STACK..`

Comment (obastemur) [inline pal/src/thread/thread.cpp]: Discard this one. In that case (if system doesn't support limits etc.) `pthread_attr_getstack` should fail.

**Comment by obastemur** [comment]
> fno-omit-frame-pointer should not noticeably affect build time (it might slow down runtime on functions with high register pressure).

I'm more optimistic with runtime performance. 

> The debug build speed up was done by improving the performance of IsAddressOnStack- it used to call GetThreadStackBounds very frequently and the Linux implementation of that function is slow

@digitalinfinity I didn't understand how that would improve debug build time? Are we using this method on build ?

LGTM

**Comment by jianchun** [comment]
> I didn't understand how that would improve debug build time?

He didn't mean speed up "build time". He meant speed up debug build binary.

**Pull request merged**